### PR TITLE
New Rule: at least business support enabled

### PR DESCRIPTION
**Pull request by bksarthak**
I confirm these files are made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)

No issue number applicable

New rule for checking if Business Support is enabled.

### CODE REVIEW

**Comment by jongogogo** [comment]
This rule just check if there is a support api available, not if it is business level. Everyone should have access to support API. Not really sure what is the logic here to determine if it is business level.

**Comment by jongogogo** [comment]
Good reference: https://github.com/awslabs/aws-config-rules/issues/151

**Comment by bksarthak** [comment]
As DescribeCases API is only allowed when Business support is enabled, this rule tests if an error is thrown when you try to use that API call. If ClientError exception is thrown it tells that business support is not enabled on the account. I've also added another exception code where it will show noncompliant if the role doesn't have enough permission to use Support API

**Comment by bksarthak** [comment]
```
'''
  #####################################
  ##           Gherkin               ##
  #####################################  
  Rule Name:
  BUSINESS_SUPPORT_ENABLED
  
  Description:
  Check whether the AWS Account has Business support plans enabled.
  
  Trigger:
  Frequency every 3hours
  
  Reports on:
  AWS::::Account
  
  Rule Parameters:
  None
  
  Scenarios:
  Scenario: 1
    Given: AWS Account has ability to use Support API's DescribeCases call with results returned without ClientError Exception
     Then: Return COMPLIANT
  Scenario: 2
    Given: AWS Support API errors out with AccessDenied exception while using the DescribeCases API call. This means the role use by Config is not allowed to use the Support API.
     Then: Return NON_COMPLIANT
  Scenario: 3
    Given: AWS Support API call errors out with SubscriptionRequired exception while using the DescribeCases API call. This error means Business Plan is not currently used for the account
     Then: Return NON_COMPLIANT
  Scenario: 4
    Given: AWS Support API call errors out with any other boto3.client exceptions
     Then: Return NOT_APPLICABLE
  
  '''
```

Adding the updated Gherkin for review (apologies I didn't know Gherkin was to be submitted first)

**Comment by jongogogo** [comment]
> 
> 
> As DescribeCases API is only allowed when Business support is enabled, this rule tests if an error is thrown when you try to use that API call. If ClientError exception is thrown it tells that business support is not enabled on the account. I've also added another exception code where it will show noncompliant if the role doesn't have enough permission to use Support API

Cool. I found the reference: https://aws.amazon.com/premiumsupport/plans/
So it means that your rule checks for Support plan Business or above (Enterprise), which is a good rule. Can you please change the Gherkin+Name of the rule to reflect that.

For your scenarios:
1) Good
2) No need, this is part of the boiler plate code
3) Good
4) No need, this mostly part of the boiler plate code. I do suggest on your `except` in your else statement to just to another `raise`.

**Comment by bksarthak** [comment]
Changed rule name to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED` on both Gherkin and Folder name.
Put a `raise` on the last `except` statement

Updated the Gherkin as follows:
```
#####################################
  ##           Gherkin               ##
  #####################################  
  Rule Name:
  BUSINESS_SUPPORT_OR_ABOVE_ENABLED
  
  Description:
  Check whether the AWS Account has Business or above(Enterprise) support plan  is enabled
  
  Trigger:
  Frequency every 3hours
  
  Reports on:
  AWS::::Account
  
  Rule Parameters:
  None
  
  Scenarios:
  Scenario: 1
    Given: AWS Account has ability to use Support API's DescribeCases call with results returned without ClientError Exception
     Then: Return COMPLIANT
  Scenario: 2
    Given: AWS Support API call errors out with SubscriptionRequired exception while using the DescribeCases API call. This error means Business Plan is not currently used for the account
     Then: Return NON_COMPLIANT
```

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
+# Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Please change the name of the file to BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+# the specific language governing permissions and limitations under the License.
+'''
+  #####################################
+  ##           Gherkin               ##
+  #####################################  
+  Rule Name:
+  BUSINESS_SUPPORT_OR_ABOVE_ENABLED
+  
+  Description:
+  Check whether the AWS Account has Business or above(Enterprise) support plan  is enabled
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Check whether the AWS Account is subscribed to the AWS Business Support Plan or above (i.e. Enterprise).

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+  
+  Trigger:
+  Frequency every 3hours
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Periodic

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+'''
+  #####################################
+  ##           Gherkin               ##
+  #####################################  
+  Rule Name:
+  BUSINESS_SUPPORT_OR_ABOVE_ENABLED
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Tab it. Idem for the rest of the Gherkin. Example: https://github.com/awslabs/aws-config-rules/blob/master/python/ELASTICSEARCH_IN_VPC_ONLY/ELASTICSEARCH_IN_VPC_ONLY.py

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    """Form the evaluation(s) to be return to Config Rules
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Remove those explanations. It is only an help for the early developers.

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    configuration_item -- the configurationItem dictionary in the invokingEvent
+    valid_rule_parameters -- the output of the evaluate_parameters() representing validated parameters of the Config Rule
+
+    Advanced Notes:
+    1 -- if a resource is deleted and generate a configuration change with ResourceDeleted status, the Boilerplate code will put a NOT_APPLICABLE on this resource automatically.
+    2 -- if a None or a list of dictionary is returned, the old evaluation(s) which are not returned in the new evaluation list are returned as NOT_APPLICABLE by the Boilerplate code
+    3 -- if None or an empty string, list or dict is returned, the Boilerplate code will put a "shadow" evaluation to feedback that the evaluation took place properly
+    """
+
+    ###############################
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Remove comment

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+    # Add your custom logic here. #
+    ###############################
+    rule_evaluations = []
+    support_client = boto3.client('support')
+    account_id = event['accountId']
+
+
+    try:
+        support_client.describe_cases()
+        annotate = 'Account has business support plan subscribed'
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: No annotation needed when it is compliant.

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+        rule_evaluations.append(build_evaluation(account_id, 'COMPLIANT', event, annotation=annotate))
+    except ClientError as error:
+        if error.response['Error']['Code'] == 'SubscriptionRequiredException':
+            annotate = 'Account does not have business support plan.'
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: This AWS Account is not subscribed to the AWS business Support plan or above.

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        if client_name == 'config':
+            return CONFIG_CLIENT_MOCK
+        elif client_name == 'sts':
+            return STS_CLIENT_MOCK
+        else:
+            raise Exception("Attempting to create an unknown client")
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('business-support-enabled')
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py]: Change to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
+import sys
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py]: Change name to: BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        elif client_name == 'sts':
+            return STS_CLIENT_MOCK
+        else:
+            raise Exception("Attempting to create an unknown client")
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('business-support-enabled')
+
+class SampleTest(unittest.TestCase):
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py]: Remove this Class. 

We need the tests for the rule though. 
`class Compliance_Test(unittest.TestCase):`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
+  "Version": "1.0", 
+  "Parameters": {
+    "CodeKey": "business-support-enabled.zip",
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json]: Change to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
+{
+  "Version": "1.0", 
+  "Parameters": {
+    "CodeKey": "business-support-enabled.zip", 
+    "SourceRuntime": "python3.6", 
+    "SourcePeriodic": "Three_Hours",
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json]: `TwentyFour_Hours`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{
+  "Version": "1.0", 
+  "Parameters": {
+    "CodeKey": "business-support-enabled.zip", 
+    "SourceRuntime": "python3.6", 
+    "SourcePeriodic": "Three_Hours", 
+    "RuleName": "business-support-enabled",
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json]: `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 2 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: changed as per suggestion

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 3 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: Done

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 8 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: done

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 7 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: removed

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 4 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: tabbed Gherkin as per given example

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 5 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: removed explanation under evaluate_compliance

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 6 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: removed comment

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py (hunk 10 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py]: changed name of the file

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py (hunk 9 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py]: changed to `BUSINESS_SUPPORT_OR_ABOVE_ENABLED`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json (hunk 13 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json]: changed to `TwentyFour_Hours`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json (hunk 12 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json]: changed

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json (hunk 14 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/parameters.json]: changed

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py (hunk 1 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled.py]: file name changed

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    rule_evaluations = []
+    support_client = boto3.client('support')
+    account_id = event['accountId']
+
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: remove extra line

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py (hunk 11 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/business-support-enabled_test.py]: remove SampleTest class and added Compliance_Test class with two test cases as per Gherkin

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py (hunk 15 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: removed extra line

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    def test_scenario_1(self):
+        resp_expected = []
+        response = RULE.lambda_handler(build_lambda_scheduled_event(), {})
+        resp_expected.append(build_expected_response('NON_COMPLIANT', '0123456789012', DEFAULT_RESOURCE_TYPE, annotation='This AWS Account is not subscribed to the AWS business Support plan or above.'))
+        assert_successful_evaluation(self, response, resp_expected) 
+
+    def test_scenario_2(self):
+        resp_expected = []
+        response = RULE.lambda_handler(build_lambda_scheduled_event(), {})
+        resp_expected.append(build_expected_response('COMPLIANT', '123456789012', DEFAULT_RESOURCE_TYPE))
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py]: You need to mock your support API call before that. Something like that to make an return:
`SUPPORT_CLIENT_MOCK.describe_cases = MagicMock(return_value={})`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('BUSINESS_SUPPORT_OR_ABOVE_ENABLED')
+
+class Compliance_Test(unittest.TestCase):
+    def test_scenario_1(self):
+        resp_expected = []
+        response = RULE.lambda_handler(build_lambda_scheduled_event(), {})
+        resp_expected.append(build_expected_response('NON_COMPLIANT', '0123456789012', DEFAULT_RESOURCE_TYPE, annotation='This AWS Account is not subscribed to the AWS business Support plan or above.'))
```
[the remark this answers is not in the document]
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py]: completed with `side_effect`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py (hunk 16 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED_test.py]: done

**Comment by jongogogo** [comment]
Please fix the error. They are all relevant.

Check those changes to fix the boilerplate. https://github.com/awslabs/aws-config-rdk/pull/158/files

**Comment by bksarthak** [comment]
> Please fix the errors. They are all relevant.
> 
> Check those changes to fix the boilerplate. https://github.com/awslabs/aws-config-rdk/pull/158/files

committed as per guidance

**Comment by jongogogo** [comment]
************* Module BUSINESS_SUPPORT_OR_ABOVE_ENABLED
BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:150:0: E0001: invalid syntax (, line 150) (syntax-error)

**Comment by bksarthak** [comment]
> ************* Module BUSINESS_SUPPORT_OR_ABOVE_ENABLED
> BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py:150:0: E0001: invalid syntax (, line 150) (syntax-error)

the commenting was off around line 150. Accidental block comment of code instead of comments.

**Comment by bksarthak** [comment]
no errors on the linting for `BUSINESS_SUPPORT_OR_ABOVE_ENABLED` rule

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+    try:
+        support_client.describe_cases()
+        rule_evaluations.append(build_evaluation(account_id, 'COMPLIANT', event))
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: do a direct `return build_evaluation(account_id, 'COMPLIANT', event)` 
remove the variable `rule_evaluations`

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+    except ClientError as error:
+        if error.response['Error']['Code'] == 'SubscriptionRequiredException':
+            annotate = 'This AWS Account is not subscribed to the AWS business Support plan or above.'
+            rule_evaluations.append(build_evaluation(account_id, 'NON_COMPLIANT', event, annotation=annotate))
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: Same, direct return

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+        else:
```
Comment (jongogogo) [inline python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: The `else` can then by removed. and the `raise` move 1 indentations upper.

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py (hunk 18 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: done

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py (hunk 19 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: done

File: python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py (hunk 20 above)
Comment (bksarthak) [reply python/BUSINESS_SUPPORT_OR_ABOVE_ENABLED/BUSINESS_SUPPORT_OR_ABOVE_ENABLED.py]: brought `raise` one level higher and removed the `else`

**Pull request merged**